sdv/single_table/copulas.py

```diff
@@ -141,3 +147,164 @@ def _sample(self, num_rows, conditions=None):
+                univariate = univariate._instance
+
+            if univariate.PARAMETRIC == copulas.univariate.ParametricType.NON_PARAMETRIC:
+                raise NonParametricError('This GaussianCopula uses non parametric distributions')
+
+        params = self._model.to_dict()
+
+        covariance = []
+        for index, row in enumerate(params['covariance'][1:]):
+            covariance.append(row[:index + 1])
+
+        params['covariance'] = covariance
+        params['univariates'] = dict(zip(params.pop('columns'), params['univariates']))
+        params['num_rows'] = self._num_rows
+
+        return flatten_dict(params)
+
+    @staticmethod
+    def _get_nearest_correlation_matrix(matrix):
+        """Find the nearest correlation matrix.
+
+        If the given matrix is not Positive Semi-definite, which means
+        that any of its eigenvalues is negative, find the nearest PSD matrix
+        by setting the negative eigenvalues to 0 and rebuilding the matrix
+        from the same eigenvectors and the modified eigenvalues.
+
+        After this, the matrix will be PSD but may not have 1s in the diagonal,
+        so the diagonal is replaced by 1s and then the PSD condition of the
+        matrix is validated again, repeating the process until the built matrix
+        contains 1s in all the diagonal and is PSD.
+
+        After 10 iterations, the last step is skipped and the current PSD matrix
+        is returned even if it does not have all 1s in the diagonal.
+
+        Insipired by: https://stackoverflow.com/a/63131250
+        """
+        eigenvalues, eigenvectors = scipy.linalg.eigh(matrix)
+        negative = eigenvalues < 0
+        identity = np.identity(len(matrix))
+
+        iterations = 0
+        while np.any(negative):
+            eigenvalues[negative] = 0
+            matrix = eigenvectors.dot(np.diag(eigenvalues)).dot(eigenvectors.T)
+            if iterations >= 10:
+                break
+
+            matrix = matrix - matrix * identity + identity
+
+            max_value = np.abs(np.abs(matrix).max())
+            if max_value > 1:
+                matrix /= max_value
+
+            eigenvalues, eigenvectors = scipy.linalg.eigh(matrix)
+            negative = eigenvalues < 0
+            iterations += 1
+
+        return matrix
+
+    @classmethod
+    def _rebuild_correlation_matrix(cls, triangular_covariance):
+        """Rebuild a valid correlation matrix from its lower half triangle.
+
+        The input of this function is a list of lists of floats of size 1, 2, 3...n-1:
+
+           [[c_{2,1}], [c_{3,1}, c_{3,2}], ..., [c_{n,1},...,c_{n,n-1}]]
+
+        Corresponding to the values from the lower half of the original correlation matrix,
+        **excluding** the diagonal.
+
+        The output is the complete correlation matrix reconstructed using the given values
+        and scaled to the :math:`[-1, 1]` range if necessary.
+
+        Args:
+            triangle_covariange (list[list[float]]):
```

Comment:
typo, triangular_covariance. Although since we are computing the correlation matrix, made it would be better to call it triangular_correlation instead